### PR TITLE
[ENH] ANM pairwise causal discovery estimator

### LINKED ISSUE #3426: Add ANM pairwise causal discovery estimator
## 1. Is your feature request related to a problem? Please describe.

pgmpy currently provides multivariate causal discovery algorithms such as PC, HillClimbSearch, and GES, but it does not provide pairwise causal direction methods. In the bivariate setting, causal direction is not identifiable from ordinary observational equivalence alone: `X -> Y` and `Y -> X` can induce the same observed association unless additional assumptions are introduced.

Additive Noise Models provide one such assumption. ANM assumes the effect can be represented as a nonlinear function of the cause plus noise independent of the cause.

## 2. Additional context

This issue proposes adding an `ANM` estimator under `pgmpy.causal_discovery`.

Expected behavior:

- subclass `_BaseCausalDiscovery` directly
- accept exactly two-column continuous data in `fit`
- fit both directions using a configurable sklearn-compatible regressor
- use `GaussianProcessRegressor` as the default nonlinear regressor
- test residual independence
- allow CI-test overrides using pgmpy's existing convention, for example `ANM(ci_test="hsic")` or `ANM(ci_test=KCI(...))`
- populate `causal_graph_`, `adjacency_matrix_`, and `direction_score_` after fitting
- define `direction_score_` as `S(Y -> X) - S(X -> Y)`, where positive means the forward column order wins
- mark categorical data unsupported via sklearn tags

Suggested files:

- `pgmpy/causal_discovery/ANM.py`
- `pgmpy/tests/test_causal_discovery/test_ANM.py`
- update `pgmpy/causal_discovery/__init__.py`

Testing should include nonlinear additive-noise synthetic data, CI-test overrides, regressor overrides, constant inputs, NaN/inf handling, insufficient samples, and reproducibility under fixed `seed`.

## 3. Paper

Hoyer et al. Nonlinear causal discovery with additive noise models.

## 4. Reference codebase

- causal-learn ANM implementation

## 5. Related

- [Pairwise causal discovery enhancement proposal](https://github.com/pgmpy/enhancement_proposals/blob/main/2_pairwise_causal_discovery/enhancement_proposal.md)
- Future PNL estimator issue

### PR DESCRIPTION
**Pull request by hanara2112**
Add the ANM class shell, export it from causal_discovery, and a couple of smoke tests. Implementation follows in follow-up commits.

**The following checklist is mandatory**

Your PR will be closed if you remove the checklist. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md)?
- x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [x ] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

If you have used AI/LLMs for any assistance, please answer the following questions. Please refer [#2622](https://github.com/pgmpy/pgmpy/pull/2622) for an example of the level of detail we expect:

- [ x] Have you reviewed our [AI usage policy](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md#ai-usage-policy)?
- [ x] Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.

- Please list the AI tool(s) used, along with the model and its version used.

claude opus 4.8

- Please describe in **detail** how and what you used AI assistance for? Please outline your whole workflow with the AI tool.
I mainly used Claude Code (Opus) as a coding assistant, but as a reviewer and implementation helper rather than something that wrote everything for me.

My workflow was to first identify the changes I wanted to make, then have it inspect the existing pgmpy codebase and conventions before suggesting an implementation. Based on that review, I accepted or rejected ideas that fit the library. It also helped with implementation, docstrings, and incorporating maintainer feedback. I reviewed every suggested change myself before committing.

- What steps have you taken to verify that the changes correctly address the issue? What edge cases have you considered? Other than running tests, what else have you verified?

Apart from the unit tests and the runnable doctest, I checked several edge cases.

The tests cover constant columns, incorrect number of variables, NaN/non-finite values, categorical inputs, custom regressors with custom CI tests, and reproducibility when using the same random_state.

For the new scoring logic, I also checked the case where a CI test doesn't provide effect_size_ (for example generalized_cov). Instead of getting an unclear None - None error, it now raises a clear ValueError.

I also verified the estimator on a simple nonlinear dataset to make sure it correctly recovered the direction X -> Y and produced a positive direction_score_.

Finally, I manually checked categorical-input rejection, verified that the doctest output still matched, scanned the files for accidental AI-style or non-ASCII characters (only "Schölkopf" remains intentionally), ensured everything stayed within the 120-character line limit, and ran the pre-commit hooks (ruff lint and formatting).

- Have you used AI for generating tests? Can you compress them into a smaller number of tests without losing coverage?

Yes. I used AI to help write the tests as well.

I also think the four input-validation tests can be simplified into a single parametrized pytest test since they're all following the same pytest.raises(ValueError) pattern. That keeps the same coverage while reducing repetition.

I'd still keep the other tests separate since they check different behavior—initialization, direction recovery, reproducibility, and custom overrides.

### Issue number(s) that this pull request fixes
- Fixes #3426 

### List of changes to the codebase in this pull request
  1. ANM.py – Added input validation, updated the scoring method to use effect_size_, improved error handling, and revised the docstrings.
  2. test_ANM.py – Added tests for input validation, reproducibility, custom regressors/CI tests, and direction recovery.
  3. estimators/__init__.py – Added the ANM estimator to the public exports.

### CODE REVIEW

**Comment by ankurankan** [comment]
@hanara2112 PR checklist needs to be a lot more detailed. Look at the example linked.

File: pgmpy/causal_discovery/ANM.py
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py:1]: Use 120 char line lenght. You should be able to change the settings in your editor.

Comment (ankurankan) [reply pgmpy/causal_discovery/ANM.py:1]: Please go through the file and make everything 120 char

File: pgmpy/causal_discovery/ANM.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        self : pgmpy.causal_discovery.ANM
+            Returns the instance with the fitted attributes.
+        """
+        if X.shape[1] != 2:
+            raise ValueError(f"ANM requires exactly two variables, got {X.shape[1]}.")
+
+        if get_dataset_type(X) != "continuous":
+            raise ValueError("ANM requires continuous (numeric) variables; got non-continuous data.")
+
+        x, y = X.columns
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py]: Use `self.feature_names_in_` instead of `X.columns`. If we change the preprocessing of feature names in the future, using `self.feature_names_in_` would make sure that this doesn't fail or silently take incorrect values.

File: pgmpy/causal_discovery/ANM.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        for col in (x, y):
+            if X[col].std() == 0:
+                raise ValueError(f"Variable '{col}' is constant; ANM requires non-constant variables.")
+
+        score_forward = self._direction_score(X[x], X[y])
+        score_backward = self._direction_score(X[y], X[x])
+        self.direction_score_ = score_backward - score_forward
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py]: I think we decided to remove this? This doesn't make sense as scores can be both positive and negative.

File: pgmpy/causal_discovery/ANM.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+        edge = (x, y) if self.direction_score_ >= 0 else (y, x)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py]: You will need to compare the absolute values of `score_forward` and `score_backward`.

File: pgmpy/causal_discovery/ANM.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        if get_dataset_type(X) != "continuous":
+            raise ValueError("ANM requires continuous (numeric) variables; got non-continuous data.")
+
+        x, y = X.columns
+        for col in (x, y):
+            if X[col].std() == 0:
+                raise ValueError(f"Variable '{col}' is constant; ANM requires non-constant variables.")
+
+        score_forward = self._direction_score(X[x], X[y])
+        score_backward = self._direction_score(X[y], X[x])
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py]: Use named arguments.

File: pgmpy/causal_discovery/ANM.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+class ANM(BaseCausalDiscovery):
+    """
+    Bivariate causal discovery using Additive Noise Models (ANM) [1]_.
+
+    Given two continuous variables, ANM orients the edge between them by assuming
+    ``effect = f(cause) + noise`` with the noise independent of the cause. It fits
+    a regressor in both directions and picks the one whose residuals are most
+    independent of the input. It always returns a direction; use
+    ``direction_score_`` to judge how confident that call is.
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py]: Should mention the assumptions of the algorithm here.

File: pgmpy/causal_discovery/ANM.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        :class:`~sklearn.gaussian_process.GaussianProcessRegressor` is used.
+
+    ci_test : str or BaseCITest instance, default=None
+        Independence test between the cause and the residuals; must expose an effect
+        size. If ``None``, uses ``"gcm"``.
+
+    random_state : int, default=None
+        Random seed for the default regressor.
+
+    Attributes
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py:43]: There are more attributes that get set in the `BaseCausalDiscovery` class. Have a look at other algorithm.

File: pgmpy/causal_discovery/ANM.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        The learned causal graph with the single oriented edge.
+
+    adjacency_matrix_ : pd.DataFrame
+        Adjacency matrix representation of ``causal_graph_``.
+
+    direction_score_ : float
+        Orientation confidence in ``[-1, 1]``: the residual-dependence effect size of
+        the reverse direction minus that of the forward direction. ``> 0`` means the
+        first column causes the second, ``< 0`` the reverse, and values near ``0``
+        mean low confidence.
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py]: Shour just expose forward and backward score.

File: pgmpy/causal_discovery/ANM.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        if self.regressor is None:
+            # RBF and Gaussian-noise GP kernel (Hoyer et al., 2009); args are (init, (lower, upper)) opt bounds.
+            regressor = GaussianProcessRegressor(
+                kernel=C(1.0, (1e-3, 1e3)) * RBF(1.0, (1e-2, 1e2)) + WhiteKernel(0.1, (1e-10, 1e1)),
+                random_state=self.random_state,
+            )
+        else:
+            regressor = clone(self.regressor)
+
+        cause_2d = cause.to_numpy().reshape(-1, 1)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py]: You can directly pass it as a 2D array where you call this method.

File: pgmpy/causal_discovery/ANM.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        regressor.fit(cause_2d, effect.to_numpy())
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py]: I don't think you need to convert it to numpy arrays. sklearn should accept pandas dataframes.

File: pgmpy/causal_discovery/ANM.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        residual = effect.to_numpy() - regressor.predict(cause_2d)
+
+        resid_data = pd.DataFrame({cause.name: cause.to_numpy(), "_residual": residual})
+        ci_test = get_ci_test(test="gcm" if self.ci_test is None else self.ci_test, data=resid_data)
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py]: Why are you passing `test="gcm"` here? The `get_ci_test` method can automatically choose a test based on data if `test=None`

File: pgmpy/causal_discovery/ANM.py [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        ci_test.run_test(cause.name, "_residual", Z=[])
+
+        if ci_test.effect_size_ is None:
```
Comment (ankurankan) [inline pgmpy/causal_discovery/ANM.py]: Which test doesn't have an `effect_size_`?

File: pgmpy/tests/test_causal_discovery/test_ANM.py
Comment (ankurankan) [inline pgmpy/tests/test_causal_discovery/test_ANM.py:1]: Tests need to check for exact scores returned by the method. This would help us avoid errors if the values returned by these methods change silently in future.

**Review by ankurankan — CHANGES_REQUESTED** [review]

File: pgmpy/causal_discovery/ANM.py (hunk 1 above)
Comment (aryamanbahl2112-arch) [reply pgmpy/causal_discovery/ANM.py]: x, y = self.feature_names_in_

File: pgmpy/causal_discovery/ANM.py (hunk 11 above)
Comment (hanara2112) [reply pgmpy/causal_discovery/ANM.py]: among the cont. test, only the generalized_cov and the edge-case independece_match. I kept it as a light safeguard, can drop as for it's narrow usecase

File: pgmpy/causal_discovery/ANM.py
Comment (hanara2112) [reply pgmpy/causal_discovery/ANM.py:1]: i checked it, the longest line is of 118 chars and nothing exceeds 120. 
is there any other way to check it or change in my ide?

File: pgmpy/causal_discovery/ANM.py [hunk 12]
```diff
@@ -148,9 +145,13 @@ def _direction_score(self, cause: pd.DataFrame, effect: pd.Series) -> float:
         """
         if self.regressor is None:
             # RBF and Gaussian-noise GP kernel (Hoyer et al., 2009); args are (init, (lower, upper)) opt bounds.
+
+            from sklearn.gaussian_process import GaussianProcessRegressor
+            from sklearn.gaussian_process.kernels import RBF, WhiteKernel
+            from sklearn.gaussian_process.kernels import ConstantKernel as C
+
```
Comment (hanara2112) [inline pgmpy/causal_discovery/ANM.py]: i thought we wanted imports at the top of the file?

Comment (ankurankan) [reply pgmpy/causal_discovery/ANM.py]: @hanara2112 I think in this case it makes sense to have it inside to avoid triggering the import when the causal discovery module is imported.

**Review by ankurankan — COMMENTED** [review]
1. Should change the argument names ` cause` and `residual` in the scoring methods.

**Review by ankurankan — APPROVED** [review]
